File: contest/LeetCode_Weekly_Contest_202/Magnetic_Force_Between_Two_Balls.cpp
class Solution {
public:
    bool check(vector<int>& p, int dist, int m) {
        int n = p.size();
        int cnt = 1;
        int pre = p.front();
        for (int i = 1; cnt < m && i < n; ++i) {
            if (p[i] - pre >= dist) {
                cnt++;
                pre = p[i];
            }
        }
        return cnt >= m;
    }
    int maxDistance(vector<int>& p, int m) {
        sort(p.begin(), p.end());
        int low = 0, hight = 1e9, ans = 0;
        while (low <= hight) {
            int mid = (low + hight) >> 1;
            if (check(p, mid, m)) {
                ans = mid;
                low = mid + 1;
            } else {
                hight = mid - 1;
            }
        }
        return ans;
    }
};

File: contest/LeetCode_Weekly_Contest_202/Magnetic_Force_Between_Two_Balls.js
/**
 * @param {number[]} position
 * @param {number} m
 * @return {number}
 */
var maxDistance = function(p, m) {
    p.sort(function(a, b) {return a-b;});
    
    var check = function(dist) {
        var cnt = 1, pre = p[0], n = p.length;
        for (var i = 1; i < n && cnt < m; ++i) {
            if (p[i] - pre >= dist) {
                pre = p[i];
                cnt++;
            }
        }
        return cnt == m;
    };
    
    var low = 0, hight = 1000000000, ans = 0;
    while (low <= hight) {
        var mid = (low + hight) >> 1;
        if (check(mid)) {
            ans = mid;
            low = mid + 1;
        } else {
            hight = mid - 1;
        }
    }
    return ans;
};
